include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -1,13 +1,89 @@
+#include <nbl/builtin/hlsl/cpp_compat.hlsl>
+#include <nbl/builtin/hlsl/concepts.hlsl>
+#include <nbl/builtin/hlsl/fft/common.hlsl>
+
 #ifndef _NBL_BUILTIN_HLSL_WORKGROUP_FFT_INCLUDED_
 #define _NBL_BUILTIN_HLSL_WORKGROUP_FFT_INCLUDED_
 
+// ------------------------------- COMMON -----------------------------------------
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace workgroup
+{
+namespace fft
+{
+
+template<uint16_t _ElementsPerInvocationLog2, uint16_t _WorkgroupSizeLog2, typename _Scalar NBL_PRIMARY_REQUIRES(_ElementsPerInvocationLog2 > 0 && _WorkgroupSizeLog2 >= 5)
+struct ConstevalParameters
+{
+    using scalar_t = _Scalar;
+
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t ElementsPerInvocationLog2 = _ElementsPerInvocationLog2;
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t WorkgroupSizeLog2 = _WorkgroupSizeLog2;
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t TotalSize = uint32_t(1) << (ElementsPerInvocationLog2 + WorkgroupSizeLog2);
+
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t ElementsPerInvocation = uint16_t(1) << ElementsPerInvocationLog2;
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t WorkgroupSize = uint16_t(1) << WorkgroupSizeLog2;
+
+    // Required size (in number of uint32_t elements) of the workgroup shared memory array needed for the FFT
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t SharedMemoryDWORDs = (sizeof(complex_t<scalar_t>) / sizeof(uint32_t)) << WorkgroupSizeLog2;
+};
+
+}
+}
+}
+} 
+// ------------------------------- END COMMON ---------------------------------------------
+
+// -------------------------------- CPP ONLY ----------------------------------------------
+
+#ifndef __HLSL_VERSION
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace workgroup
+{
+namespace fft
+{
+
+struct OptimalFFTParameters
+{
+    uint16_t elementsPerInvocationLog2;
+    uint16_t workgroupSizeLog2;
+};
+
+inline OptimalFFTParameters optimalFFTParameters(const uint32_t maxWorkgroupSize, uint32_t inputArrayLength)
+{
+    // This is the logic found in core::roundUpToPoT to get the log2
+    const uint16_t workgroupSizeLog2 = 1u + findMSB(min(inputArrayLength / 2, maxWorkgroupSize) - 1u);
+    const uint16_t elementsPerInvocationLog2 = 1u + findMSB(max((inputArrayLength >> workgroupSizeLog2) - 1u, 1u));
```

Comment:
There's an issue `inputArrayLength==2` leads to `findMSB(0)`, which I'm not sure if both our C++ and HLSL impl return `-1` in that case (c.f. with @Przemog1 )

You should also write an assert before computation of `elementsPerInvocationLog2` that `(inputArrayLength>>workgroupSizeLog2)>1` (obviously commented out for HLSL)

There should be no `max(,1)` necessary, as workgroup size may never be larger than input array length.

TBH I'd first **round up** the inputArrayLength to PoT and **round down** the maxWorkgroupSize to PoT while working out their logarithms, then proceed with the rest of the computations